### PR TITLE
Single head attention, decoupled LR, autoregressive auxiliary loss, and gradient accumulation

### PR DESCRIPTION
**Pull request by lucidrains**
clean PR

### CODE REVIEW

File: bonito/models/configs/dna_r9.4.1@v3.1.toml [hunk 1]
```diff
@@ -34,6 +34,7 @@ heads = 6
 attn_dropout = 0.1
 ff_dropout = 0.1
 use_self_attn = false
+num_encoder_layers_attend = 6
```
Comment (lucidrains) [inline bonito/models/configs/dna_r9.4.1@v3.1.toml:37]: 6 because it also includes the fmap right after the conv encoder (and before the first RNN)

File: bonito/crf/model.py [hunk 2]
```diff
@@ -135,6 +135,23 @@ def ctc_viterbi_alignments(self, scores, targets, target_lengths):
         return viterbi_alignments(stay_scores, move_scores, target_lengths + 1 - self.state_len)
 
 
+class Encoder(nn.Module):
```
Comment (samgd) [inline bonito/crf/model.py:138]: missing `from torch import nn`

Comment (lucidrains) [reply bonito/crf/model.py]: thanks!

File: bonito/models/configs/dna_r9.4.1@v3.1.toml [hunk 3]
```diff
@@ -30,6 +30,8 @@ num_attn_heads = 1
 use_isab_attn = true
 isab_num_latents = 6
 
+weight_tie_attn_blocks = false
```
Comment (lucidrains) [inline bonito/models/configs/dna_r9.4.1@v3.1.toml:33]: for parameter saving when using ISAB blocks, which has twice the number of attention parameters than S(M)HA blocks

File: bonito/models/configs/dna_r9.4.1@v3.1.toml [hunk 4]
```diff
@@ -22,5 +22,27 @@ rnn_type = "lstm"
 activation = "swish"
 blank_score = 2.0
 
+attn_layers = [ 3, 4 ]    # the list of layers (1 - 5) after which to play an attention block, right after the LSTM layer
+attn_dropout = 0.1        # attention dropout, keep at 0.1
+ff_dropout = 0.1          # feedforward dropout, keep at 0.1
+num_attn_heads = 1        # number of attention heads, which should be kept at 1 for single-head attention, but can be increased to > 1 to turn on multi-head attention
+dim_attn_head = 64        # dimension per attention head, should just keep at 64, but can be lowered to 32 for further efficiency / perf tradeoff
+
+use_isab_attn = false     # whether to use ISAB attention (induced-set attention block from the Set Transformers paper)
```
Comment (lucidrains) [inline bonito/models/configs/dna_r9.4.1@v3.1.toml:31]: if you were to set this to true, the number of attention heads need to be increased to 4 or above. a good starting config would be

```
num_attn_heads = 4
dim_attn_head = 64
use_isab_attn = true
isab_num_latents = 6
```
